Try: Improve Stack and Row setup states.

jasmussen opened the pull request:
## What?

This PR takes a small step towards improving Row and Stack setup states. Before, there was no visible indication of what the Row or Stack blocks were meant for, and block spacing and padding did not affect the setup state:
![before](https://user-images.githubusercontent.com/1204802/161501862-33ebcf7e-31e0-45e0-b702-801e62c8cee2.gif)

After, dashed outlines indicate direction and purpose, and spacing and padding affect even the setup state:

![after](https://user-images.githubusercontent.com/1204802/161501870-3693259e-792b-4012-a48a-8efbf97571a2.gif)

Note that this PR does not touch the vanilla flow based Group variation, nor is it meant to affect any other blocks that use the Button Block appender.

## Why?

The Row and Stack setup states are in a bad place. There's zero indication that these blocks are in fact flex-based containers affected by block spacing controls and other flex behaviors, or that the Row block is horizontal in direction. 

## How?

A dashed line is output to indicate where a second block could sit, giving more of an indication of what the block is all about. 

If this one can land, a next step could potentially be to introduce a dedicated placeholder state, a la this:

<img width="1640" alt="Group" src="https://user-images.githubusercontent.com/1204802/161025388-3f972bde-9ef7-4560-a703-4cea1f7ec067.png">
<img width="1640" alt="Row" src="https://user-images.githubusercontent.com/1204802/161025394-c7ab96a1-8d1f-480b-a80b-941d7c1940f3.png">

<img width="1640" alt="Stack" src="https://user-images.githubusercontent.com/1204802/161025403-ab6b4bf1-c63a-44c8-be47-fc50a9f7f3df.png">

Normally a placeholder like that is something to be careful with introducing, and introducing mainly for content-first blocks, otherwise they have a risk of generating layout shifts.

The group is a little unique in that vein, since through transforms and grouping tools such as [this multi-select toolbar transform](#39920), it seems likely that you'll be using the group as a side-efffect of those tools. Meaning, when you insert the block directly, we can afford to show more context.

## Testing Instructions

Test Group, Row and Stack blocks.

**Note:** this PR isn't shippable as is. It's a little hacky, but most importantly it affects all appenders not just those of the Row and Stack blocks. I would appreciate help here: can we add a CSS class to the Row and Stack blocks, even if just editor-only, so we can scope the extra rules?

jasmussen committed:
Try: Improve Stack and Row setup states.

jasmussen commented:
CC: @youknowriad @andrewserong @glendaviesnz as I think you've worked on both row and stack blocks. Thank you for looking.

youknowriad commented on packages/block-editor/src/components/button-block-appender/style.scss:
```diff
@@ -3,12 +3,16 @@
+	// Needs specificity to override button styles.
+	&.components-button.components-button {
+		padding: $grid-unit-15;
+	}
```
Note that this component impacts more than just the "group" block. It can be used by third-party blocks... Just want to make sure the change is safe for everyone.

jasmussen replied:
The problem right now is that the padding registered for this button doesn't take at all, because it's overridden by `.edit-post-visual-editor .components-button`:
<img width="440" alt="Screenshot 2022-03-31 at 14 41 19" src="https://user-images.githubusercontent.com/1204802/161057038-c42b5988-2eb8-4530-a776-e9186c357398.png">

So this change does increase the padding to `12px`, but shipping in trunk it's currently `6px 12px` because of the inheritance.

youknowriad commented on packages/block-editor/src/components/default-block-appender/style.scss:
```diff
@@ -112,8 +112,35 @@
+		// Let the parent be selectable in the placeholder area.
+		pointer-events: none;
+		.block-editor-inserter {
+			pointer-events: all;
```
Can you clarify a bit how these styles changes work :) and if they are safe for all blocks.

jasmussen replied:
Yes indeed! So, one of the nice things about the dashed boxes this PR adds is that, at least for the Row and Stack blocks, it gives you a "safe" area to select the empty block, where right now for the base group block, the whole button takes up the space so you can't really select the block itself.

Only, this didn't work in the initial implementation, because clicking that dashed outline would select `.block-list-appender`, instead of `.block-editor-block-list__block` (which would let you select the block):
![row](https://user-images.githubusercontent.com/1204802/161058188-6c3751ce-8d70-4445-9c04-b2e10eff9267.gif)

By disabling pointer events for the placeholder stuff, then re-enabling it on the appender inside, we get the best of both worlds. I dislike touching pointer-events generally, but in this case it was the only way I could see this working.

youknowriad reviewed (commented):
I see the dotted block for the group block as well but block gap doesn't seem to be taken into consideration there. Is this intended?

I really like this PR btw

jasmussen commented:
> I see the dotted block for the group block as well but block gap doesn't seem to be taken into consideration there. Is this intended?

Thank you for looking!

Yes, this is why it's a draft. I feel like the styles here can't be really successful unless we tailor them a little bit to the three block variants. I tried my best with what's there, but without some CSS classes to ambiguate between the three variations, I don't think I can get these styles to a "safe" place. Would it be terrible to have a CSS class even if in the editor only, so I can tweak the styles better?

youknowriad commented:
>  Would it be terrible to have a CSS class even if in the editor only, so I can tweak the styles better?

I think it's fine in the editor. Even the frontend it might be fine but we need to be careful what to do there and we should do it in a unified/consistent way, so for now safer to start with just editor.

Also, if CSS is not enough `renderAppender` can also be tweaked in JS.

jasmussen commented:
> I think it's fine in the editor. Even the frontend it might be fine but we need to be careful what to do there and we should do it in a unified/consistent way, so for now safer to start with just editor.

I think that'd be totally fine.

> Also, if CSS is not enough renderAppender can also be tweaked in JS.

I think there's actually an opportunity here, but I don't know how big of a task it is. Right now this appender component, is used for both the fixed-position black plus, and the "empty block" white/bordered appender. From last I looked at this, it caused the code to be a little messy in places, since the two behave quite differently, but are named the same. I think there could be a big code cleanup here if the component was optimized for that split. 

You are free to push to this PR, by the way, if you like and have time!

paaljoachim commented:
Thank you for creating this PR, Joen!

Testing:

### Before
The Group block and Row block.
![Screenshot 2022-04-01 at 10 54 01](https://user-images.githubusercontent.com/5323259/161231263-046c2e0d-489c-4d28-b1f5-6696b73afc0b.jpg)

Selecting the Row block:
![Screenshot 2022-04-01 at 10 55 07](https://user-images.githubusercontent.com/5323259/161231296-c96fe0c9-4715-45dd-828d-e548a605139a.jpg)


### After
![Screenshot 2022-04-01 at 10 58 33](https://user-images.githubusercontent.com/5323259/161231338-ed02533c-9f3b-47e7-89fd-2c3bb57deb80.jpg)


Before:
No intuitive way of clicking the Group block to select it.
No clear method for how Row block works.

After:
We are finally able to easily click to select the Group block!
Adding the outline to the right in the Row block gives a visual of what the Row block does.

This is very helpful! Thank you!
(I did not even know about the Stack block. I assume it works similar to a Group block. Only that the focus is on stacking for mobile.)

jasmussen committed:
Add editor-only flex-class to affect only row and stack.

This employs code from #39181, props @carolinan.

jasmussen commented:
Alright, I have added a CSS class in the editor only version of the Group block and optimized the CSS code to only affect the only-child initial setup states of stack and row. That makes this ready. Before:
![before](https://user-images.githubusercontent.com/1204802/161501862-33ebcf7e-31e0-45e0-b702-801e62c8cee2.gif)

After:

![after](https://user-images.githubusercontent.com/1204802/161501870-3693259e-792b-4012-a48a-8efbf97571a2.gif)

I have added the CSS class based on code and conversation in #39181 (thank you and props @carolinan), so in the editor there's now a `is-layout-flex` class for Row and Stack blocks. This PR does not touch the frontend, and I'm happy to revisit this code if the approach needs more time to mature.

jasmussen commented:
> We are finally able to easily click to select the Group block!

I'm afraid that was a temporary side effect of this PR, and not the intended solution. However I acknowledge that's still a desirable outcome, and indeed one that would be fixed if we add a dedicated placeholder, which is something we can explore separately:

<img width="1640" alt="Group" src="https://user-images.githubusercontent.com/1204802/161502692-22aa0074-991e-4a1e-bf7a-ffa21ac52bcf.png">

jasmussen commented:
There's a failing test on this one and I can't figure out if it's legitimate or not. I've restarted it once more, but otherwise I'd love advice on it, in addition to a review, perhaps a ✅ if you have one? After the latest change this one feels very safe to me.

youknowriad commented on packages/block-library/src/group/edit.js:
```diff
@@ -52,7 +57,14 @@ function GroupEdit( { attributes, setAttributes, clientId } ) {
 
-	const blockProps = useBlockProps();
+	const classes = classnames( className, {
+		'is-layout-flex': type === 'flex',
```
Should this be a `is-layout-{ type }` for consistency?

jasmussen replied:
I guess it would accomplish the same, right? How would I write the left side of that statement, though? 🙈

youknowriad replied:
no need, also no need to use the "className" prop, I think it's added automatically already. So basically, this can just be:


```
const blockProps = useBlockProps( {
 	className: `is-layout-${ type }`,
 } );
```

jasmussen replied:
Nice one. Pushed!

jasmussen committed:
Change classname.

Props @youknowriad

jasmussen committed:
Restore old rule.

youknowriad reviewed (approved):
LGTM 👍
